plugin-niassa/src/main/java/ca/on/oicr/gsi/shesmu/niassa/MigrationAction.java

```diff
@@ -0,0 +1,287 @@
+        limsKeysCollection
+            .signatures()
+            .collect(
+                Collectors.groupingBy(
+                    p -> new SimpleLimsKey(p.first()),
+                    Collectors.mapping(Pair::second, Collectors.toSet())));
+
+    try (AutoCloseable timer = matchTime.start(Long.toString(workflowAccession))) {
+      matches =
+          workflowAccessions()
+              .distinct()
+              .boxed()
+              .flatMap(
+                  accession ->
+                      server
+                          .get()
+                          .analysisCache()
+                          .get(accession)
+                          .map(
+                              as ->
+                                  as.compare(
+                                      workflowAccessions(),
+                                      Long.toString(majorOliveVersion),
+                                      fileMatchingPolicy,
+                                      inputFileSWIDs,
+                                      limsKeys,
+                                      signatures,
+                                      annotations)))
+              .filter(
+                  pair ->
+                      pair.comparison() != AnalysisComparison.DIFFERENT
+                          && pair.actionState() == ActionState.SUCCEEDED)
+              .sorted()
+              .collect(Collectors.toList());
+    } catch (InitialCachePopulationException e) {
+      final Metadata metadata = server.get().metadata();
+      final Workflow workflow = metadata.getWorkflow((int) workflowAccession);
+      metadata.clean_up();
+      if (workflow == null) {
+        this.errors =
+            Collections.singletonList("Niassa doesn't seem to recognise this workflow SWID.");
+        cacheCollision = false;
+        return ActionState.FAILED;
+      }
+      this.errors =
+          Collections.singletonList(
+              "Another workflow is fetching the previous workflow runs from Niassa.");
+      cacheCollision = true;
+      return ActionState.WAITING;
+    } catch (Exception e) {
+      e.printStackTrace();
+      return ActionState.FAILED;
+    }
+
+    if (matches.isEmpty()) {
+      // go into waiting state and try again later
+      this.errors = Collections.singletonList("No candidate workflow runs for migration.");
+      return ActionState.WAITING;
+    } else {
+      ObjectMapper MAPPER = new ObjectMapper();
+      ObjectNode labels = MAPPER.createObjectNode();
+      annotations.forEach(labels::put);
+
+      request.setConsumableResources(new TreeMap<>());
+      request.setTarget("MIGRATION");
+      request.setWorkflow(workflowName);
+      request.setWorkflowVersion(matches.get(0).state().workflowVersion());
+      request.setLabels(labels);
+
+      if (stale) {
+        return ActionState.ZOMBIE;
+      }
+      final var throttled = actionServices.isOverloaded(this.services);
```

Comment:
This can be dropped. We're not using the services for real, so it doesn't matter.